python/pecos/slr/gen_codes/gen_qasm.py

```diff
@@ -291,14 +292,22 @@ def qgate_qasm(self, op, repr_str: str | None = None):
                 op.qargs = (op.qargs,)
 
         for q in op.qargs:
-            if isinstance(q, tuple):
+            if isinstance(q, QReg):
+                # Broadcasting across a qubit register is inconsistent with the current Permute
+                # strategy, so "unroll" broadcast operations to make them act on individual qubits.
+                # See, for example, https://github.com/PECOS-packages/PECOS/issues/95.
+                lines = [f"{repr_str} {qubit};" for qubit in q]
+                str_list.extend(lines)
+
+            elif isinstance(q, tuple):
                 if len(q) != op.qsize:
                     msg = f"Expected size {op.qsize} got size {len(q)}"
                     raise Exception(msg)
                 qs = ",".join([str(qi) for qi in q])
                 str_list.append(f"{repr_str} {qs};")
 
-            str_list.append(f"{repr_str} {str(q)};")
+            else:
+                str_list.append(f"{repr_str} {q};")
```

Comment:
I'm pretty sure(?) that the unconditional final `str_list.append(...)` was a bug, and should have been inside an `else` statement.  Otherwise this method first acts the gate on every `qs in q` and then again appends a gate on (the tuple) `q`.

Reply:
To me, SWAP is a unitary gate that could possibly be converted to a physical permute if a compiler what smart enough. There is a need to distinguish the two operations as you may what fine control and choose one over the other and do not want to rely on a compiler.

As far as rewriting QASM, the user should no longer be responsible with debugging that. This is the job of the language (e.g., SLR)/compiler. If the underlying base language had support for Permute, then that would make things cleaner; however, SLR needs to metaprogram features that are not supported by the base language which is essentially what higher languages have to do when converting to lower languages. They provide more facilities then the simpler underlying language provides.

That being said. I want to rewrite the Permute object to be a little more friendly to code generation and just stores the info in a nice manner to be handed off to the code generator function.

Reply:
Upgrading SWAP to a "native gate" of the hardware is perfectly compatible with thinking of it as a unitary gate.  Relabeling qubits **_does_** implement the desired unitary operation, and it is an **advantage** of trapped ion platform that it can perform SWAP gates perfectly 🙂!  Other platforms can't do this!  This is also a capability that can be leveraged by circuit-level compilers (search [this paper](https://arxiv.org/pdf/2309.05157) for "swap mirroring", for example), but doing so requires that the hardware recognize SWAP as a native gate.  This essentially "exposes" the virtual SWAP capability to compilers that work on the level of (say) `pytket`, `cirq`, or `qiskit`.

There is also precedent here: having a virtual native SWAP gate is directly analogous to virtual Rz gates on IBM devices.  IBM devices accept Rz instructions, but do not actually apply any physical operations any qubits when asked to "apply" an Rz gate (they instead make a software update of a rotating frame).

If you want to "decompose" a SWAP into physical gates (say, to test those gates), you can write a circuit with (say) 3 CNOTs (or rather, 3 MS gates), and run that circuit in "verbatim" mode to bypass a compiler.

In any case, what are your thoughts of unrolling broadcast operations in this PR?